pkg/report/genericsbom.go

```diff
@@ -0,0 +1,141 @@
+package report
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/package-url/packageurl-go"
+	"golang.org/x/xerrors"
+)
+
+type GsbomPackage struct {
+	Purl         string   `json:"purl,omitempty"`
+	Relationship string   `json:"relationship,omitempty"`
+	Dependencies []string `json:"dependencies,omitempty"`
+}
+
+type GsbomFile struct {
+	SrcLocation string `json:"source_location,omitempty"`
+}
+
+type GsbomManifest struct {
+	Name string     `json:"name,omitempty"`
+	File *GsbomFile `json:"file,omitempty"`
+	//TODO can also be number or boolean
+	Metadata map[string]string       `json:"metadata,omitempty"`
+	Resolved map[string]GsbomPackage `json:"resolved,omitempty"`
+}
+
+type GsbomJob struct {
+	Name string `json:"name,omitempty"`
+	Id   int    `json:"id,omitempty"`
+}
+
+type Gsbom struct {
+	Version   int                      `json:"version,omitempty"`
+	Detector  string                   `json:"detector,omitempty"`
+	Ref       string                   `json:"ref,omitempty"`
+	Sha       string                   `json:"sha,omitempty"`
+	Job       *GsbomJob                `json:"job,omitempty"`
+	Scanned   string                   `json:"scanned,omitempty"`
+	Manifests map[string]GsbomManifest `json:"manifests,omitempty"`
+}
+
+type GsbomWriter struct {
+	Output io.Writer
+}
+type depsRslvr func(result Result) map[string]GsbomPackage
+
+func (gsbmw GsbomWriter) Write(report Report) error {
+	gsbom := &Gsbom{}
+
+	gsbom.Scanned = time.Now().Format(time.RFC3339)
+	gsbom.Detector = "trivy"
+
+	//TODO optionally add git information
+	manifests := make(map[string]GsbomManifest)
+
+	for _, result := range report.Results {
+		manifest := GsbomManifest{}
+		manifest.Name = result.Type
+		//show path for languages only
+		if result.Class == ClassLangPkg {
+			manifest.File = &GsbomFile{
+				SrcLocation: result.Target,
+			}
+		}
+		var resolver depsRslvr
+		if result.Packages != nil {
+			resolver = resolvedFromPackages
+		} else {
+			resolver = resolvedFromVulns
+		}
+		manifest.Resolved = resolver(result)
+		manifests[result.Target] = manifest
+	}
+
+	gsbom.Manifests = manifests
+
+	output, err := json.MarshalIndent(gsbom, "", "  ")
+	if err != nil {
+		return xerrors.Errorf("failed to marshal generic sbom: %w", err)
+	}
+
+	if _, err = fmt.Fprint(gsbmw.Output, string(output)); err != nil {
+		return xerrors.Errorf("failed to write generic sbom: %w", err)
+	}
+	return nil
+}
+func buildPurl(pkgName, version, pkgType string) string {
```

Comment:
https://github.com/package-url/purl-spec#purl

The GitHub user or organization should be set as the namespace.
Therefore, @ and other characters present in the test case will be escaped.

Reply:
I read the link but it's not clear how to extract github user/organization from generic package. I mean there are many types of packages for some there is no obvious way and for some it's simple